util_for_test.go

```diff
@@ -0,0 +1,329 @@
+
+	// DBForTest.Prepare is to run any SurrealQL for test prep. This would fail
+	// early if the provided string results in an error from SurrealDB server.
+	db.Prepare(t, "CREATE user:x SET name = 'x'")
+
+	// More interactions can simply use db instance from here on.
+}
+
+func Test_TestSurrealDBParallelSimple(t *testing.T) {
+	t.Parallel()
+	instanceCount := 20
+	for i := 0; i < instanceCount; i++ {
+		t.Run(fmt.Sprintf("parallel__%d", i), func(t *testing.T) {
+			t.Parallel()
+
+			_, db, close := NewTestSurrealDB(t)
+			defer close()
+
+			// Ensure that CREATE statement does not run against the same
+			// database (which would fail).
+			db.Prepare(t, "CREATE user:x SET name = 'x';")
+		})
+	}
+}
+
+func Test_TestSurrealDBParalleSelect(t *testing.T) {
+	t.Parallel()
+	instanceCount := 20
+	for i := 0; i < instanceCount; i++ {
+		i := i
+		t.Run(fmt.Sprintf("parallel__%d", i), func(t *testing.T) {
+			t.Parallel()
+			_, db, close := NewTestSurrealDB(t)
+			defer close()
+
+			// Ensure that CREATE statement does not run against the same
+			// database (which would fail).
+			db.Prepare(t, fmt.Sprintf("CREATE user:x SET name = 'x', index = %d;", i))
+
+			// Query the saved data.
+			data, err := db.Select("user")
+			if err != nil {
+				t.Fatalf("failed to select: %v", err)
+			}
+			jsonBytes, err := json.Marshal(data)
+			if err != nil {
+				t.Fatalf("failed to marshal to json: %v", err)
+			}
+			// Note the Index field, which would change based on the for loop.
+			type dataholder struct {
+				ID    string `json:"id,omitempty"`
+				Name  string `json:"name,omitempty"`
+				Index int    `json:"index,omitempty"`
+			}
+			d := []dataholder{}
+			err = json.Unmarshal(jsonBytes, &d)
+			if err != nil {
+				t.Fatalf("failed to unmarshal data: %v", err)
+			}
+			// Simply check all fields match as expected.
+			want := dataholder{ID: "user:x", Name: "x", Index: i}
+			if d[0].ID != want.ID || d[0].Name != want.Name || d[0].Index != want.Index {
+				t.Fatalf("data mismatch:\n    want: %+v, got:  %+v", want, d)
+			}
+		})
+	}
+}
+
+type DBForTest struct {
+	*DB
+}
+
+func NewTestSurrealDB(t testing.TB) (string, *DBForTest, func()) {
+	t.Helper()
+
+	checkSurrealCLIOnce.Do(func() {
+		if _, err := exec.LookPath("surreal"); err == nil {
+			testHasSurrealCLI = true
+		}
+	})
+	if !testHasSurrealCLI {
+		t.Skip("surreal CLI not found, skipping")
+	}
+
+	port := getFreePort(t)
+	srvEndpoint := fmt.Sprintf("0.0.0.0:%d", port)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, "surreal",
+		"start",
+		"--auth",
+		"--user", "root",
+		"--pass", "surrealdb",
+		"--bind", srvEndpoint,
+		"memory")
+	// Ensure to wait for a short while before context cancellation to
+	// propagate, so that the client can shut down before.
+	// NOTE: There is currently a race condition and potential panic due to the
+	// closed websocket being used in a tight loop.
+	cmd.WaitDelay = DelayBeforeServerExit
```

Comment:
Indeed, this has to do with the websocket connection handling, and this logic should not be necessary. But without this, tests would fail sporadically.